### DISCUSSION TITLE
Revisiting final control statements in `switch` statements

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#6578] [Language Ideas]
## Summary

Switch statements in C# follow the C syntax, but impose a restriction to avoid code bugs, you *need* to specify a control statement before exiting a case label in the switch statement. This was done to prevent accidental falls to the next label, or exit out of the statement.

There are two problems with this; this approach completely disables implicit falling under the next label, and enforces usage of `goto`. Using `goto` in itself is not as detrimental, but things get complicated when you add C# syntax sugar to the mix. Switch cases like `case 3 when x = 4` are not reachable unless you add a custom label after the `case` label, which is a minor inconvenience.

While making complex case labels more accessible is not exactly easily achievable, something can be done about falling under the next label, which can be enabled through the use of another keyword, something like `fall`. A `fall;` statement will simply move to the next case label within the switch statement.

Another improvement I would like to see is disabling the need for a `break;` statement at the end of a case label to get out of the switch statement. With this proposal, falling to the next statement is enabled through another keyword and is not considered implied behavior. This can also be supported by the fact that switch case labels are already special labels that break the statement down into multiple sections. Normal labels have no special meaning for the compiler, and therefore are not getting behavioral adjustments from this proposal.

## Examples

With all of the above in mind, consider the following example:
```csharp
public void ApplyAction(ActionType actionType, IUser user)
{
    switch (actionType)
    {
        case ActionType.DeleteCascade:
            // first cascade delete all entries related to the user
            goto case ActionType.Delete; // then perform the same actions related to deleting a user
        case ActionType.Delete:
            // delete the user from the db and perform the deactivation operations
            // as if the user is not an admin, regardless of their role
            goto deactivateNonAdmin;
        case ActionType.Deactivate when user is not IAdminUser:
        deactivateNonAdmin:
            // deactivate non-admin user
            break;
        case ActionType.Deactivate when user is IAdminUser:
        deactivateAdmin:
            // deactivate admin user
            break;
    }
}
```

Which can be rewritten as follows:
```csharp
public void ApplyAction(ActionType actionType, IUser user)
{
    switch (actionType)
    {
        case ActionType.DeleteCascade:
            // first cascade delete all entries related to the user
            fall; // then perform the same actions related to deleting a user
        case ActionType.Delete:
            // delete the user from the db and perform the deactivation operations
            // as if the user is not an admin, regardless of their role
            fall;
        case ActionType.Deactivate when user is not IAdminUser:
        deactivateNonAdmin:
            // deactivate non-admin user
        case ActionType.Deactivate when user is IAdminUser:
        deactivateAdmin:
            // deactivate admin user
    }
}
```

**Comment by HaloFour** [comment]
Something like this should fall out naturally from the combination of [switch expressions as statements](https://github.com/dotnet/csharplang/issues/2632) and [sequence expressions](https://github.com/dotnet/csharplang/issues/377).  By building the ergonomics into the newer form of `switch` it avoids the problems with ambiguity and confusion that would arise from attempting to redesign the existing C# `switch` statement, which is extremely unlikely to be considered by the language team.

**Reply by Rekkonnect** [reply]
I cannot see this combination of features completely replacing the standard switch statement. Yes, you do not use switch statements in your everyday life now that you have switch expressions, but it is not replaceable in all circumstances. And in this case it's not completely redesigning it, but rather getting rid of some annoyances that are well-known that cause users to avoid using it.

**Reply by HaloFour** [reply]
With that combination of features you could use switch expressions in nearly all cases where you use switch statements today, and the alternative syntax avoids the problems that arise with case fallthrough.  The existing syntax is not going to be changed, so this is the best route to address those "annoyances".

**Comment by rbwhitaker** [comment]
It's a little too bad you couldn't use the `continue` keyword to mean "fall through to the next thing" to keep the number of keywords lower. Alas, adding that particular thing in now could break a `continue` that actually belongs to an outer loop. I can't imagine that's all that common, but it is syntactically valid today, and I'm sure _somebody_ out there is using it.

**Reply by Rekkonnect** [reply]
I'm personally using it in my code and it's nearly guaranteed that it'll break tons of code. `continue` makes sense and should be applicable, if it was designed like that from day 1.

**Reply by CyrusNajmabadi** [reply]
> I can't imagine that's all that common, but it is syntactically valid today, and I'm sure somebody out there is using it.

Roslyn is certainly using it.  It's totally normal to be looping over things, but then switch over things in teh middle.  Think of bog standard stuff like lexing :)

**Reply by rbwhitaker** [reply]
Fair points. :D And to clarify, a loop containing a switch _is_ commonplace in the work I do. What I hadn't expected to be commonplace was a loop containing a switch containing a continue. Apparently, that's more common than I had expected!
